### PR TITLE
[Jenkins 53257] client deletes

### PR DESCRIPTION
**Pull request by mandie722**
[JENKINS-53257](https://issues.jenkins-ci.org/browse/JENKINS-53257)

Client will delete plugins even if there's no updates

### CODE REVIEW

File: distribution/client/test/helpers.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ * tests
+ */
+const fs            = require('fs');
+
+class Helpers {
+  constructor () {
+  }
+
+  checkFileExists(filename) {
+    return new Promise(r=>fs.access(filename, fs.F_OK, e => r(!e)));
```
[the remark this answers is not in the document]
Comment (batmat) [reply distribution/client/test/helpers.js]: Why not `fs.existsSync(path)`?
https://nodejs.org/api/fs.html#fs_fs_existssync_path

Also do you really want to be async in that case? I guess not in tests?

File: distribution/client/src/lib/update.js [hunk 2]
```diff
@@ -74,7 +74,7 @@ class Update {
         updates.core.checksum.signature));
     }
 
-    if ((!updates.plugins) || (!updates.plugins.updates)) {
+    if ((!updates.plugins) || (!updates.plugins.updates && !updates.plugins.deletes)) {
       logger.debug('No plugins available in the response');
       this.updateInProgress = null;
       this.saveUpdateSync(updates);
```
Comment (batmat) [inline distribution/client/src/lib/update.js]: :fire_engine: 
Unrelated with your change, but isn't there a huge bug here? If only the core is updated, we'll bail out without restart nor snapshotting, but the war has been downloaded and put at the location ready isn't it?

File: distribution/client/src/lib/update.js [hunk 3]
```diff
@@ -86,13 +86,19 @@ class Update {
...
+        logger.info('Downloading', plugin);
+        tasks.push(Downloader.download(plugin.url,
+          Storage.pluginsDirectory(),
+          `${plugin.artifactId}.hpi`,
+          plugin.checksum.signature));
+      });
+    }
+
+    if (updates.plugins.deletes) {
+      tasks.push(Storage.removePlugins(updates.plugins.deletes));
```
Comment (batmat) [inline distribution/client/src/lib/update.js]: Can we add more `logger.info` and the likes inside those `if` conditions? I feel at this stage of the project, we want rather more logs than fewer don't we?

File: distribution/client/src/lib/storage.js [hunk 4]
```diff
@@ -59,6 +61,21 @@ class Storage {
...
+    }
+    let pluginPath = this.pluginsDirectory();
+    let retArray = [];
+    plugins.forEach((plugin) => {
+      retArray.push(fs.unlink(`${pluginPath}/${plugin}.hpi`, () => {
+        logger.info(`${pluginPath}/${plugin}.hpi was deleted`);
+        UI.publish(`Deleted ${plugin}.hpi`);
+      }));
+    });
+    return retArray;
```
[the remark this answers is not in the document]
Comment (mandie722) [reply distribution/client/src/lib/storage.js]: I have tests that confirm this works, but they conflict with the snapshotting changes, so I'm still rewriting to get it all working together.

File: distribution/client/test/helpers.js [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+  }
+
+  touchFile(filename) {
```
Comment (batmat) [inline distribution/client/test/helpers.js]: sync `fs.openSync`, or even `fs.writeFileSync('/path/to/theFile', 'blah content')` might be easier for the purpose?

File: distribution/client/test/storage.test.js [hunk 6]
```diff
@@ -29,4 +42,26 @@ describe('The storage module', () => {
...
+      const pluginPath = Storage.pluginsDirectory();
+      mkdirp.sync(pluginPath);
+      filenames.forEach((filename) => {
+        h.touchFile(`${pluginPath}/${filename}.hpi`);
+        expect(h.checkFileExists(`${pluginPath}/${filename}.hpi`)).resolves.toBeTruthy();
+      });
+      await Storage.removePlugins(filenames);
+      filenames.forEach((filename) => {
+        expect(h.checkFileExists(`${pluginPath}/${filename}.hpi`)).resolves.toBeFalsy();
+      });
```
Comment (batmat) [inline distribution/client/test/storage.test.js]: nice one

File: distribution/client/src/lib/update.js (hunk 2 above)
Comment (mandie722) [reply distribution/client/src/lib/update.js]: true, hadn't found that one yet, we have a severe lack of positive testing that I'm trying to fix

**Pull request merged**